mlkem/ntt.c

```diff
@@ -9,137 +9,402 @@
-  ensures(array_abs_bound(r, 0, start + 2*len - 1, bound + MLKEM_Q))
-  ensures(array_abs_bound(r, start + 2 * len, MLKEM_N - 1, bound)))
+  requires(memory_no_alias(r, sizeof(pc)))
+  requires(array_abs_bound(r, 0, MLKEM_N - 1, NTT_BOUND1))
+  assigns(memory_slice(r, sizeof(pc)))
+  ensures(array_abs_bound(r, 0, MLKEM_N - 1, NTT_BOUND4)))
 {
-  /* `bound` is a ghost variable only needed in the CBMC specification */
   int j;
-  ((void)bound);
-  for (j = start; j < start + len; j++)
+  for (j = 0; j < 32; j++)
   __loop__(
-    invariant(start <= j && j <= start + len)
-    /*
-     * Coefficients are updated in strided pairs, so the bounds for the
-     * intermediate states alternate twice between the old and new bound
-     */
-    invariant(array_abs_bound(r, 0,           j - 1,           bound + MLKEM_Q))
-    invariant(array_abs_bound(r, j,           start + len - 1, bound))
-    invariant(array_abs_bound(r, start + len, j + len - 1,     bound + MLKEM_Q))
-    invariant(array_abs_bound(r, j + len,     MLKEM_N - 1,     bound)))
+    /* A single iteration of this loop updates 8 coefficients 3 times each,
+     * meaning their bound jumps from NTT_BOUND1 to NTT_BOUND4. Other (as yet
+     * untouched) coefficients remain bounded by NTT_BOUND1. When this loop
+     * terminates with j == 32, ALL the coefficients have been updated
+     * exactly 3 times, so ALL are bounded by NTT_BOUND4, which establishes
+     * the post-condition */
+    invariant(0 <= j && j <= 32)
+    invariant(array_abs_bound(r,       0,   j - 1, NTT_BOUND4))
+    invariant(array_abs_bound(r,       j,      31, NTT_BOUND1))
+    invariant(array_abs_bound(r,      32,  j + 31, NTT_BOUND4))
+    invariant(array_abs_bound(r,  j + 32,      63, NTT_BOUND1))
+    invariant(array_abs_bound(r,      64,  j + 63, NTT_BOUND4))
+    invariant(array_abs_bound(r,  j + 64,      95, NTT_BOUND1))
+    invariant(array_abs_bound(r,      96,  j + 95, NTT_BOUND4))
+    invariant(array_abs_bound(r,  j + 96,     127, NTT_BOUND1))
+    invariant(array_abs_bound(r,     128, j + 127, NTT_BOUND4))
+    invariant(array_abs_bound(r, j + 128,     159, NTT_BOUND1))
+    invariant(array_abs_bound(r,     160, j + 159, NTT_BOUND4))
+    invariant(array_abs_bound(r, j + 160,     191, NTT_BOUND1))
+    invariant(array_abs_bound(r,     192, j + 191, NTT_BOUND4))
+    invariant(array_abs_bound(r, j + 192,     223, NTT_BOUND1))
+    invariant(array_abs_bound(r,     224, j + 223, NTT_BOUND4))
+    invariant(array_abs_bound(r, j + 224, (MLKEM_N - 1), NTT_BOUND1)))
   {
-    int16_t t;
-    t = fqmul(r[j + len], zeta);
-    r[j + len] = r[j] - t;
-    r[j] = r[j] + t;
+    const int ci1 = j + 0;
+    const int ci2 = j + 32;
+    const int ci3 = j + 64;
+    const int ci4 = j + 96;
+    const int ci5 = j + 128;
+    const int ci6 = j + 160;
+    const int ci7 = j + 192;
+    const int ci8 = j + 224;
+    int16_t t1, t2;
+
+    /* Layer 1 */
+    t1 = fqmul(r[ci5], mlkem_l1zeta1);
```

Comment:
This may be more readable if you introduced either a macro or a static inline function for a single butterfly.

Reply:
That sounds like a good idea. I will try.

Reply:
Please see #611